File: src/main/java/com/labback/data/domain/groups/StudyGroup.java
package com.labback.data.domain.groups;

import jakarta.persistence.*;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "study_group")
public class StudyGroup {
    private static final String SEQ_NAME = "study_groups_id_seq";

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = SEQ_NAME)
    @SequenceGenerator(name = SEQ_NAME, sequenceName = SEQ_NAME, allocationSize = 1)
    private Long id;

    @NotNull
    @NotBlank
    private String name;

    @NotNull
    @OneToOne
    @JoinColumn(name = "coordinates_id")
    private Coordinates coordinates;

    private String creationDate;

    @Min(value = 1, message = "Students count must be greater than 0")
    @NotNull
    private Integer studentsCount;

    @Enumerated(EnumType.STRING)
    private FormOfEducation formOfEducation;

    @NotNull
    @Enumerated(EnumType.STRING)
    private Semester semesterEnum;

    @OneToOne
    @JoinColumn(name = "admin_id")
    private Person groupAdmin;

    private String owner;
}

File: src/jsMain/kotlin/data/groups/StudyGroup.kt
package data.groups


class StudyGroup {
    constructor(id: Long,
                name: String,
                coordinates: Coordinates,
                creationDate: String,
                studentsCount: Int,
                formOfEducation: FormOfEducation?,
                semesterEnum: Semester,
                groupAdmin: Person,
                owner: String
        ) {
        this.id = id
        this.name = name
        this.coordinates = coordinates
        this.creationDate = creationDate
        this.studentsCount = studentsCount
        this.formOfEducation = formOfEducation
        this.semesterEnum = semesterEnum
        this.groupAdmin = groupAdmin
        this.owner = owner

    }

    var id: Long? = null
        set(value) {
            if (value != null) {
                field = value
            }
        }
    var name: String? = null
        set(value) {
            if (value != null) {
                field = value
            }
        }
    var coordinates: Coordinates? = null
        set(value) {
            if (value != null) {
                field = value
            }
        }
    var creationDate: String? = null
        set(value) {
            if (value != null) {
                field = value
            }
        }
    var studentsCount: Int? = null
        set(value) {
            if (value != null && value > 0) {
                field = value
            }
        }
    var formOfEducation: FormOfEducation? = null
    var semesterEnum: Semester? = null
        set(value) {
            if (value != null) {
                field = value
            }
        }
    var groupAdmin: Person? = null
        set(value) {
            if (value != null) {
                field = value
            }
        }
    var owner: String? = null
        set(value) {
            if (value != null) {
                field = value
            }
        }

    fun getJson(): String {
        return "{\"id\":${id},\"name\":\"${name}\"," +
                "\"coordinates\":{\"x\":${coordinates!!.x}, \"y\":${coordinates!!.y}}," +
                "\"creationDate\":null,\"studentsCount\":${studentsCount}," +
                "\"formOfEducation\":${if (formOfEducation == null) "null" else "\"${formOfEducation}\""}," +
                "\"semesterEnum\":\"${semesterEnum}\"," +
                "\"groupAdmin\": {\"name\":\"${groupAdmin!!.name}\", \"weight\":${groupAdmin!!.weight}," +
                "\"eyeColor\":\"${groupAdmin!!.eyeColor}\", \"location\":{" +
                "\"x\":${groupAdmin!!.location!!.x}, \"y\":${groupAdmin!!.location!!.y}" +
                ", \"z\":${groupAdmin!!.location!!.z}}}}"
    }
}